Unify all <position> / <bg-position> around strong CSS value representation

weinig opened the pull request:
#### 25aee767cc7bb6f3445ed6b9356292f58f42d731
<pre>
Unify all &lt;position&gt; / &lt;bg-position&gt; around strong CSS value representation
<a href="https://bugs.webkit.org/show_bug.cgi?id=291960">https://bugs.webkit.org/show_bug.cgi?id=291960</a>

Reviewed by Darin Adler.

- Removes older &lt;position&gt; / &lt;bg-position&gt; parsing code, replacing it with
  the strong value parser wrapped in new CSSPositionValue, CSSPositionXValue
  and CSSPositionYValue CSSValue derived classes.
- Simplifies background-position style representation to be consistent with
  other position usages, removing the origin fields and using the calc(100% - x)
  for opposite edge instead.
- Updates serialization of computed value of background-position to match
  the spec, always serializing the offset as a pair of &lt;length-position&gt; values.

* LayoutTests/fast/css/background-position-serialize-expected.txt:
* LayoutTests/fast/css/background-position-serialize.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-x-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-y-computed-expected.txt:
* LayoutTests/transitions/background-position-transitions-expected.txt:
* LayoutTests/transitions/background-position-transitions.html:
    - Update tests and results for updated background-position serialization.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPositionValue.cpp: Added.
* Source/WebCore/css/CSSPositionValue.h: Added.
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
    - Add new CSSPositionValue derived classes.

* Source/WebCore/css/CSSToStyleMap.cpp:
    - Update to use strong value conversion.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
    - Update for removal of explicit origin.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
    - Update to call strong value parser, using new split operator as needed.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h:
    - Replace old non-strong value parsers with strong value equivalents.

* Source/WebCore/css/values/primitives/CSSPosition.cpp:
* Source/WebCore/css/values/primitives/CSSPosition.h:
    - Add support for 3-component variant used by &lt;bg-position&gt;.
    - Adds a new split operator for taking a CSS::Position and getting
      a CSS::PositionX and CSS::PositionY out of it. Used by shorthand
      parsers.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
    - Update for new name, TwoComponentPositionHorizontalVertical.

* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/style/FillLayer.cpp:
* Source/WebCore/rendering/style/FillLayer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
    - Remove explicit originX/originY for background position, instead
      using calc(100% - x) for opposite edge cases (matching other
      position usages).

* Source/WebCore/style/StyleBuilderConverter.h:
    - Simplify conversion to call strong value conversion code.

* Source/WebCore/style/values/primitives/StylePosition.cpp:
* Source/WebCore/style/values/primitives/StylePosition.h:
    - Refactor conversion code to more easily handle more cases, adding in
      the 3-component variant.

* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
    - Fix conversion issue where zoom was not applied to calc() used
      in &lt;length-percentage&gt; values. We now always use canonicalize(),
      which is much nicer.

Canonical link: <a href="https://commits.webkit.org/294038@main">https://commits.webkit.org/294038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c50c0e07cd3cefd2be4f1eb1843b43352170573

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100664 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20316 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10615 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105801 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51252 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102705 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20624 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28790 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76657 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33690 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15833 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90925 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57011 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/15648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8932 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50628 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9007 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108156 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27782 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28145 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87126 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/85156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21657 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29850 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7578 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21770 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27717 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/32970 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27528 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30846 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29086 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSPositionValue.cpp:
```diff
@@ -0,0 +1,117 @@
+}
+
+
+} // namespace WebCore
```
```suggestion
}

} // namespace WebCore
```

darinadler commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -1898,6 +1898,13 @@ anchors-valid
+// right
+// top
+// bottom
+// center
```
Is this important?

weinig replied:
No. It can go.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -469,34 +469,19 @@ static inline Ref<CSSValue> valueForReflection(const StyleReflection* reflection
 }
 
-static Ref<CSSValueList> createPositionListForLayer(CSSPropertyID propertyID, const FillLayer& layer, const RenderStyle& style)
+static Ref<CSSValueList> createPositionListForLayer(CSSPropertyID, const FillLayer& layer, const RenderStyle& style)
```
```suggestion
static Ref<CSSValueList> createPositionListForLayer(const FillLayer& layer, const RenderStyle& style)
```

weinig committed:
Unify all <position> / <bg-position> around strong CSS value representation
https://bugs.webkit.org/show_bug.cgi?id=291960

Reviewed by Darin Adler.

- Removes older <position> / <bg-position> parsing code, replacing it with
  the strong value parser wrapped in new CSSPositionValue, CSSPositionXValue
  and CSSPositionYValue CSSValue derived classes.
- Simplifies background-position style representation to be consistent with
  other position usages, removing the origin fields and using the calc(100% - x)
  for opposite edge instead.
- Updates serialization of computed value of background-position to match
  the spec, always serializing the offset as a pair of <length-position> values.

* LayoutTests/fast/css/background-position-serialize-expected.txt:
* LayoutTests/fast/css/background-position-serialize.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-x-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-y-computed-expected.txt:
* LayoutTests/transitions/background-position-transitions-expected.txt:
* LayoutTests/transitions/background-position-transitions.html:
    - Update tests and results for updated background-position serialization.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPositionValue.cpp: Added.
* Source/WebCore/css/CSSPositionValue.h: Added.
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
    - Add new CSSPositionValue derived classes.

* Source/WebCore/css/CSSToStyleMap.cpp:
    - Update to use strong value conversion.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
    - Update for removal of explicit origin.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
    - Update to call strong value parser, using new split operator as needed.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h:
    - Replace old non-strong value parsers with strong value equivalents.

* Source/WebCore/css/values/primitives/CSSPosition.cpp:
* Source/WebCore/css/values/primitives/CSSPosition.h:
    - Add support for 3-component variant used by <bg-position>.
    - Adds a new split operator for taking a CSS::Position and getting
      a CSS::PositionX and CSS::PositionY out of it. Used by shorthand
      parsers.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
    - Update for new name, TwoComponentPositionHorizontalVertical.

* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/style/FillLayer.cpp:
* Source/WebCore/rendering/style/FillLayer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
    - Remove explicit originX/originY for background position, instead
      using calc(100% - x) for opposite edge cases (matching other
      position usages).

* Source/WebCore/style/StyleBuilderConverter.h:
    - Simplify conversion to call strong value conversion code.

* Source/WebCore/style/values/primitives/StylePosition.cpp:
* Source/WebCore/style/values/primitives/StylePosition.h:
    - Refactor conversion code to more easily handle more cases, adding in
      the 3-component variant.

* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
    - Fix conversion issue where zoom was not applied to calc() used
      in <length-percentage> values. We now always use canonicalize(),
      which is much nicer.

Canonical link: https://commits.webkit.org/294038@main
